WIP: Stream Permissions

krtkmj opened the pull request:
Backend to give permissions to streams users.

smarx commented:
*Automated message from Dropbox CLA bot*

@krtkmj, it looks like you've already signed the Dropbox CLA. Thanks!

krtkmj commented:
@neerajwahi Need comments before I move on to give stream administration rights to stream owner.

timabbott commented:
I'd like us to use a model where a stream can have more than one owner.  So you should instead make this a property of the Subscription model with a name like `is_owner`, and write a convenience function to fetch the owners of a stream.

krtkmj commented:
@timabbott: In my last meeting with @acrefoot and @neerajwahi, we decided that in invite only streams we can have only one stream owner. When an invite only stream is created, the user who is creating the stream will be `stream_owner` by default. If stream_owner of the stream wishes to leave, then he/she will have to reassign anyone else from the stream as new stream_owner.

And, since stream_owner is directly related to stream. Why should this be a property of Subscription Model? What can be possible problems in this design?

neerajwahi commented on zerver/tests/test_subs.py:
```diff
@@ -402,6 +402,36 @@ def test_admin_remove_others_from_unsubbed_private_stream(self):
+    def test_create_invite_only_stream_owner(self):
+        # type: () -> None
+        """
+        When a user creates a invite only stream, create him/her
```
Awkward phrasing. I'd make it something like "A user who creates an invite only stream should by default be the owner of that stream."

neerajwahi commented on static/templates/subscription_type.handlebars:
```diff
@@ -1,2 +1,5 @@
-{{#if invite_only}}{{t 'This is an <span class="icon-vector-lock"></span> <b>invite-only stream</b>. Only people who have been invited can access its content, but&nbsp;any&nbsp;member&nbsp;of&nbsp;the&nbsp;stream can&nbsp;invite&nbsp;others.' }}
-{{else}}{{t 'This is a <span class="icon-vector-globe"></span> <b>public stream</b>. Anybody in your organization can join.' }}{{/if}}
+{{#if invite_only}}
+	{{t 'This is an <span class="icon-vector-lock"></span> <b>invite-only stream</b>' }}. This is owned by {{ stream_owner }}. {{t 'Only people who have been invited can access its content, but&nbsp;any&nbsp;member&nbsp;of&nbsp;the&nbsp;stream can&nbsp;invite&nbsp;others.' }}
```
This isn't working. I'm just seeing the string "This is owned by " in my dev VM with no owner specified.

neerajwahi commented:
@timabbott: @acrefoot and I told @krtkmj to assume at most one owner of a stream, but if you'd like to support a multiple owner model we should go with that (and maybe rename 'owner' to 'admin' since owner might imply just one)

@krtkmj could you please modify the diff to handle multiple owners as @timabbott mentions above?

timabbott commented:
Yeah, I think we should support multiple owners/admins.  Part of my reasoning is that it's a useful feature to support more than one; part of it is that this will make the code paths for this much more similar to create other user roles on streams.  `stream administrator` is a fine user-facing name for that role; I guess we can call it `stream_admin` in the code; I'd prefer we never abbreviate it as simply `admin` to avoid confusion with realm administrations (which are `realm_admin` in the code).

krtkmj commented:
@neerajwahi I have updated this PR. Please check if there is any flaw in current implementation.

neerajwahi commented on static/js/subs.js:
```diff
@@ -3,10 +3,13 @@ var subs = (function () {
 var exports = {};
 
 function add_admin_options(sub) {
+    var extend_dict = {};
```
This variable doesn't seem to be used

neerajwahi commented on zerver/migrations/0023_subscription_flags.py:
```diff
@@ -0,0 +1,20 @@
+        migrations.AddField(
+            model_name='subscription',
+            name='flags',
+            field=bitfield.models.BitField([b'stream_admin', b'moderator', b'regular', b'read_only'], default=0),
```
We should call this something more descriptive than `flags`.

Also, we may want to model this as sets of permissions instead of the user's role. So, for instance, instead of a `read_only` user role, we would have permissions like `can_write`, `can_read`, `can_moderate`, `can_add_remove_users`, etc... @timabbott  thoughts?

timabbott replied:
Yeah, I agree, we should do this based on permissions in the style you describe @neerajwahi, rather than roles, at least one the backend, since we'll probably eventually want to enable users to define sets of permissions that don't form a single hierarchy.  

One thing that this does point out to me is that we'll probably need a "default user permissions" field to store what permissions users who subscribe to the stream (or are subscribed by another user) get on joining.

krtkmj commented:
@neerajwahi, @timabbott : Rendering permissions to frontend gives an integer. Converting this integer to binary gives permission bits.
And, while I was coding this, I tried to add a few more permissions(like splitting `can_add_remove_users` to `can_add_users` and `can_remove_users`). I had to check everywhere and update these permission checks. In long term, this can be hard for maintenance and [django-bitfield](https://github.com/disqus/django-bitfield) doesn't have proper documentation on this. Any suggestion or anything I am missing something?

neerajwahi commented on static/js/compose.js:
```diff
@@ -56,6 +56,16 @@ function autosize_textarea() {
+    var can_write = _.find(stream.permissions, function (permission){
+        return permission === "can_write";
+    });
+    if (can_write === undefined && tabname === "stream") {
```
Why would `can_write` be `undefined` here if the user doesn't have permission to write to the stream? From the anonymous function above it looks like it should be either `true` or `false`

krtkmj replied:
`_.find` returns `undefined` if no value passes the test. http://underscorejs.org/#find

neerajwahi replied:
Ah ok. Better to use `_.some` then because it's confusing if a variable called `can_write` is not a boolean

neerajwahi commented on static/js/subs.js:
```diff
@@ -3,10 +3,16 @@ var subs = (function () {
 var exports = {};
 
 function add_admin_options(sub) {
+    if (sub.subscribed == true) {
```
Use `===`

neerajwahi commented on static/js/subs.js:
```diff
@@ -3,10 +3,16 @@ var subs = (function () {
+        var can_moderate_stream = page_params.is_admin || can_moderate;
+        var can_add_remove_users = _.contains(sub.permissions, 'can_add_remove_users');
+    }
     return _.extend(sub, {
```
The codepath below is hit even when `sub.subcribed` is false, in which case `can_moderate_stream` is undefined

neerajwahi commented on static/js/subs.js:
```diff
@@ -3,10 +3,16 @@ var subs = (function () {
-        'is_admin': page_params.is_admin,
-        'can_make_public': page_params.is_admin && sub.invite_only && stream_data.is_subscribed(sub.name),
-        'can_make_private': page_params.is_admin && !sub.invite_only
+        'is_admin': can_moderate_stream,
```
This code is hit even when `sub.subcribed` is false, in which case `can_moderate_stream` is undefined

krtkmj replied:
Fixed.

neerajwahi commented on static/js/subs.js:
```diff
@@ -3,10 +3,16 @@ var subs = (function () {
 
 function add_admin_options(sub) {
+    if (sub.subscribed == true) {
+        var can_moderate = _.contains(sub.permissions, 'can_moderate');
```
Nit: you don't really need the `can_moderate` variable. I'd probably just put the expression on the right inline in the assignment of `can_moderate_stream` below

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -1298,10 +1327,16 @@ def bulk_add_subscriptions(streams, users):
     subs_to_add = [] # type: List[Tuple[Subscription, Stream]]
     for (user_profile, recipient_id, stream) in new_subs:
         color = pick_color_helper(user_profile, subs_by_user[user_profile.id])
+        # Give stream admin permission is stream is created by user.
```
Typo: s/is/if

neerajwahi commented on zerver/lib/actions.py:
```diff
@@ -1298,10 +1327,16 @@ def bulk_add_subscriptions(streams, users):
         color = pick_color_helper(user_profile, subs_by_user[user_profile.id])
+        # Give stream admin permission is stream is created by user.
+        if current_user == user_profile:
+            permissions = 15
```
Avoid magic numbers. Replace the right side of the number with an explicit construction. Take a look at https://github.com/zulip/zulip/blob/master/zerver/models.py#L1198 for how the UserMessage model handles flags

neerajwahi commented on zerver/models.py:
```diff
@@ -571,6 +571,11 @@ class Stream(ModelReprMixin, models.Model):
     description = models.CharField(max_length=1024, default=u'') # type: text_type
 
+    # Specify default permissions assigned when user subscribes to a stream.
+    PERMISSION_FLAGS = ['can_read', 'can_write', 'can_moderate',
```
Since `Stream` and `Subscription` both share the same flags, you probably only want one `PERMISSION_FLAGS` variable that's shared between both classes.

Some options off the top of my head include making `PERMISSION_FLAGS` global, or keeping it in one class (for example `Stream`) and referring to the class variable from `Subscription`. The second option seems better to me

krtkmj replied:
Fixed.

neerajwahi commented on zerver/models.py:
```diff
@@ -1278,7 +1289,11 @@ class Subscription(ModelReprMixin, models.Model):
+                        'can_add_remove_users']
+    permissions = BitField(flags=PERMISSION_FLAGS) # type: BitHandler
+
+    DEFAULT_STREAM_COLOR = "#c2c2c2"
```
Why did you remove the Unicode designation (`u...`) from the `DEFAULT_STREAM_COLOR` constant?

krtkmj replied:
Sorry, didn't notice this while rebasing with master.

neerajwahi commented on zerver/migrations/0023_auto_20160724_1657.py:
```diff
@@ -0,0 +1,25 @@
+        migrations.AddField(
+            model_name='stream',
+            name='default_permissions',
+            field=bitfield.models.BitField([b'can_read', b'can_write', b'can_moderate', b'can_add_remove_users'], default=1),
```
Setting `default` to 1 is confusing because it's not immediately clear which element in the list that's referring to (do we read the list from the left or the right?). Better to replace `default=1` with `default=[b'can_read']`

`django-bitfield` supports a variety of ways to pass in `default`: https://github.com/disqus/django-bitfield/blob/master/bitfield/models.py#L122

neerajwahi commented on zerver/models.py:
```diff
@@ -571,6 +571,11 @@ class Stream(ModelReprMixin, models.Model):
+    # Specify default permissions assigned when user subscribes to a stream.
+    PERMISSION_FLAGS = ['can_read', 'can_write', 'can_moderate',
+                        'can_add_remove_users']
+    default_permissions = BitField(flags=PERMISSION_FLAGS, default=1) # type: BitHandler
```
See comment about `default=1` above

neerajwahi commented on zerver/tests/test_subs.py:
```diff
@@ -416,6 +416,22 @@ def test_admin_remove_others_from_unsubbed_private_stream(self):
+        stream, _ = create_stream_if_needed(user_profile.realm, stream_name)
+        bulk_add_subscriptions([stream], [user_profile], current_user=user_profile)
+        sub = get_subscription(stream_name, user_profile)
+        self.assertEqual(len(sub.permissions_list()), 4)
```
It's probably better to explicitly check that each permission in `Stream.PERMISSION_FLAGS` is satisfied here. I can imagine a situation where new permissions are added in the future (and so 4 needs to be incremented), but a bug exists that causes this test to pass anyway.

neerajwahi commented on zerver/views/streams.py:
```diff
@@ -208,12 +209,14 @@ def json_remove_subscriptions(request, user_profile):
+                                 permissions = REQ(validator=check_list(check_string), default=None)):
+    # type: (HttpRequest, UserProfile, Iterable[text_type], Optional[Iterable[text_type]], Optional[Iterable[text_type]]) -> HttpResponse
 
+    can_add_remove_users = 'can_add_remove_users' in permissions
```
Nit: might be clearer (and more consistent) to replace this with
```
can_add_remove_users = user_profile.is_realm_admin or 'can_add_remove_users` in permissions
```

neerajwahi commented:
Thanks for the PR @krtkmj! Looks pretty good -- made some inline comments. I'm going to run this in the dev vm now and take a look at the interactions.

Also, looks like tests are failing due to a migration conflict.

neerajwahi commented:
@krtkmj Just tried this out and there are a bunch of issues.

- No permissions show up on existing streams:
![2016-08-02-000455_876x93_scrot](https://cloud.githubusercontent.com/assets/4666213/17319867/7bd6a3b6-5845-11e6-9335-9032848ac60c.png)

- The visual formatting of permissions could be better:
![2016-08-02-000439_336x40_scrot](https://cloud.githubusercontent.com/assets/4666213/17319882/8df8943c-5845-11e6-8ae3-7e181bdfb21c.png)

- Iago (as a realm administrator) should show up as having all permissions:
![2016-08-02-000550_804x145_scrot](https://cloud.githubusercontent.com/assets/4666213/17319907/a6308a46-5845-11e6-8a53-d051d2eef2ae.png)

- The `Administrator settings` section shows up for users with `can_moderate` permission, but it seems that they cannot actually perform any actions:
![2016-08-02-000811_881x412_scrot](https://cloud.githubusercontent.com/assets/4666213/17319964/e9ae7120-5845-11e6-9f65-cdfe3c55a957.png)

krtkmj commented:
@neerajwahi permission column shows which permissions do user have in a subscription. Since realm administrator have universal admin access which is not specific to a stream. So, rather just simply listing permission I will suggest why not show his permission as "Realm admin". Thoughts?

krtkmj commented:
@neerajwahi I updated this PR and addressed all your queries(except visual formatting of permissions). I needed suggestion on few things:
- how should UI for changing existing user's permission should look like? I had something like [this](http://www.w3schools.com/tags/tryit.asp?filename=tryhtml_select_multiple) in my mind.
- Shall stream moderator be allowed to edit message of other users?
- If stream admin changes default permissions of stream, then all existing subscribed users permissions should be updated?

neerajwahi commented:
@krtkmj I think it's fine to show the stream-level permissions for Realm admins as well. @timabbott thoughts?

timabbott commented:
Yep, that's fine, as long as it's not an invite-only stream that the realm admin isn't subscribed to.

neerajwahi commented:
@krtkmj thanks for updating! Taking a look. Does this update take care of the issues noted above during manual testing?

Re: your questions:
- I'd probably use checkboxes since you might want to add or remove multiple permissions from a user
- No, the moderator can only delete messages
- In that case we may want to ask whether or not to update existing user permissions, but we shouldn't do so automatically

neerajwahi commented:
@krtkmj please fix the failing tests.

zulipbot-test commented:
Hello @krtkmj, a Zulip maintainer reviewed this pull request over a week ago, but you haven't updated your pull request since. Please take a look at the requested changes and update your pull request accordingly.

Thank you for your valuable contributions to Zulip!

timabbott commented:
Closing this, since while it's a useful reference for doing this feature, there's enough merge conflicts that we'll probably want to type it in again before merging.

Thanks for all your work on this @krtkmj!
